[CORE-8525] - Iceberg schema evolution

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR implements the Iceberg side of Schema Evolution. In broad strokes, it adds:
1. A visitor class for traversing (and annotating) two input schemas in lockstep to determine whether one is backward compatible with the other (i.e. A is evolvable into B), structurally.
2. A small, optional, mutable, variant metadata field in `iceberg::nested_field`, intended to support (1), both in progress and for post-traversal validation.
3. A visitor class for `iceberg::nested_field::evolution_metadata`, which validates the result of (1).
4. Lots of unit tests.

The visitor pattern in (1) is largely inspired by Apache Iceberg: [CheckCompatibility.java](https://github.com/apache/iceberg/blob/apache-iceberg-1.7.1/api/src/main/java/org/apache/iceberg/types/CheckCompatibility.java) and [TypeUtil.java](https://github.com/apache/iceberg/blob/apache-iceberg-1.7.1/api/src/main/java/org/apache/iceberg/types/TypeUtil.java), though our implementation is somewhat simpler in the sense that we have `std::variant` w/ proper (enough) dynamic dispatch to arbitrary callables.

## Microbenchmark

The benchmark included with this PR seeks to compare two mechanisms for traversing a pair of input schemas and updating field IDs:
1.  "Basic" assignment which performs a simultaneous, non-recursive DFS over both schemas (copied from current `catalog_schema_manager::fill_field_ids`)
2. Annotate + Verify, which performs a single recursive DFs over both schemas, in lockstep, followed by a non-recursive DFS over the destination schema to validate & apply changes.

Notably this benchmark does not cover cases where the two schemas differ. This is primarily to test the computational cost and memory pressure of the two traversal mechanisms.

### Conclusions

- As expected, (2) takes more CPU time than (1), but generally keeps w/in the same order of magnitude.
   - Unsurprisingly, the gap increases with both nesting level and total number of fields. It is simply more work.
- The annotation step does not allocate any dynamic memory.
- Since there is an increased cost, the onus is on us to perform these traversals only when absolutely necessary.

| test                                            | iterations | median    | mad       | min       | max       | allocs | tasks | inst |     |
| ----------------------------------------------- | ---------- | --------- | --------- | --------- | --------- | ------ | ----- | ---- | --- |
| StructVisitation.BasicFieldAssign               | 928951     | 368.580ns | 0.675ns   | 367.234ns | 369.255ns | 6.000  | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransform              | 1018128    | 279.476ns | 0.636ns   | 278.840ns | 281.513ns | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransform                 | 755588     | 358.204ns | 0.368ns   | 357.526ns | 360.669ns | 14.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_5_1     | 1503843    | 191.487ns | 0.060ns   | 191.142ns | 191.900ns | 12.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_5_1    | 1714150    | 110.832ns | 0.250ns   | 110.476ns | 111.108ns | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_5_1       | 1411552    | 150.257ns | 0.158ns   | 150.099ns | 150.801ns | 6.000  | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_10_1    | 774890     | 348.358ns | 0.029ns   | 348.328ns | 349.848ns | 22.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_10_1   | 866611     | 204.199ns | 0.106ns   | 204.029ns | 204.343ns | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_10_1      | 706771     | 252.356ns | 0.004ns   | 252.346ns | 252.441ns | 11.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_20_1    | 380216     | 665.858ns | 0.078ns   | 665.654ns | 665.936ns | 42.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_20_1   | 414367     | 415.410ns | 0.078ns   | 415.288ns | 415.489ns | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_20_1      | 353225     | 464.909ns | 0.052ns   | 464.836ns | 465.052ns | 21.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_5_5     | 514812     | 536.770ns | 0.104ns   | 536.666ns | 536.968ns | 10.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_5_5    | 511058     | 559.132ns | 0.696ns   | 555.391ns | 559.829ns | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_5_5       | 420292     | 447.914ns | 0.271ns   | 447.088ns | 448.384ns | 10.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_10_5    | 237045     | 1.006us   | 0.227ns   | 1.006us   | 1.007us   | 12.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_10_5   | 228472     | 1.188us   | 1.240ns   | 1.179us   | 1.190us   | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_10_5      | 193793     | 764.561ns | 0.106ns   | 763.729ns | 764.784ns | 11.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_20_5    | 120174     | 1.895us   | 0.113ns   | 1.894us   | 1.897us   | 14.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_20_5   | 112604     | 2.465us   | 0.226ns   | 2.441us   | 2.465us   | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_20_5      | 97074      | 1.381us   | 0.059ns   | 1.380us   | 1.381us   | 12.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_100_5   | 23741      | 8.744us   | 3.199ns   | 8.738us   | 8.790us   | 18.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_100_5  | 22006      | 12.011us  | 1.029ns   | 12.008us  | 12.013us  | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_100_5     | 19258      | 6.220us   | 1.268ns   | 6.218us   | 6.221us   | 14.000 | 0.000 | 0.0  |     |
| StructVisitation.BasicFieldAssignNested_100_20  | 6426       | 37.448us  | 325.334ns | 36.492us  | 37.839us  | 18.000 | 0.000 | 0.0  |     |
| StructVisitation.DescribeTransformNested_100_20 | 4653       | 94.026us  | 54.727ns  | 93.971us  | 94.241us  | 0.000  | 0.000 | 0.0  |     |
| StructVisitation.ApplyTransformNested_100_20    | 4144       | 26.626us  | 186.807ns | 26.090us  | 28.064us  | 31.000 | 0.000 | 0.0  |     |

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
iceberg: Introduce compatibility_types module

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg: Introduce evolution_metadata

variant:
  - bool is_new
  - bool removed
  - struct src_info
    - id of the source field
    - requiredness of the source field
    - optionally source field type (if primitive)

Used in subsequent commits to:
  - cheaply identify struct fields that have changed shape
  - defer type & requiredness invariant checks to the end

Only track primitive types because
a) they are cheap to copy and
b) structured types are effectively checked during traversal

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg: Supporting types for schema evolution

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg: Introduce compatibility_utils

Some functions to support recursive iteration and transformation of the fields
under a field or struct.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/tests: Some helpers for compatibility_test

And lightly refactor test suite base class for reuse

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/compat: Introduce annotate_schema_transform

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/compat: Introduce validate_schema_transform

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/compat: Introduce evolve_schema

Make a copy of the destination schema.
Generate a schema transform from source to dest.
Apply the transform to the copy and return the result (or an error code)

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/compat: Microbenchmark for evolve_schema & friends

Generates increasingly nested structs and measures the time
it takes to:

- collect all fields of two copies w/ field_collecting_visitor
- generate a schema transform from one copy to another
- visit a generated schema transform

Serves as a decent point of comparison between the limited schema
verification already present in the tree and the utilities introduced
in this commit stack.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg: Introduce reverse_const_field_collecting_visitor

Templatize the visitor to add a const nested_field* version.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/tests: Test cases for struct evolution.

Mix of compatible & incompatible, structural & type oriented

Does not include any real tests

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
iceberg/compat: Drop unused dependencies & includes

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
CI Failures:
- [bazel-build](https://buildkite.com/redpanda/redpanda/builds/60327#01943e09-d9a1-472b-9d1e-a1dd7d953c55) failed to checkout vtools?

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/e0afd1fbb0f2c2bf9284b33e0274d92098524cc0..02b5d2c1c8652e12ca958fd9374df97505525bbd) - refactored interfaces to indicate at each step whether anything changed (fields added/removed, types promoted, requiredness changed). Might want more granularity at some point, but basically as a first step want to signal to calling code whether a metadata update is required.

oleiman commented:
CI Failures:
- [build-debug](https://buildkite.com/redpanda/redpanda/builds/60330#01943e94-cbb5-4dd1-9044-eb4b3688b85f) - failed to remove builder permission denied...infra

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6b0013f0-5820-4699-a1b8-63b232fadf69-->
<!--TEST RESULTS-->
<!--559d3ce9-f1bc-4b0c-ac81-13e1e9d103a0-->
<!--TEST RESULTS-->
<!--a7a4bb5b-5401-4939-8b4f-778795c2b980-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60330">build#60330</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60330#01943e94-cbb4-4e99-9963-4c53628a6f41 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60369">build#60369</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.consumer_offsets_consistency_test.ConsumerOffsetsConsistencyTest.test_flipping_leadership | ducktape | https://buildkite.com/redpanda/redpanda/builds/60369#01944294-7ef3-468e-8fe2-616c23b81086 | FAIL | 0/1 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60369#01944294-7ef1-4efc-9263-0be02b3b0b70 | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60379">build#60379</a></summary>


| test_id                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/60379#01944325-d6fd-4784-96eb-98185fdd5ebd | FLAKY         | 1/2      |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/60379#01944325-d6fd-4c2d-83ad-69cfb60097bb | FLAKY         | 1/2      |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60379#01944383-4eaa-4edf-9a42-07985b65aaa1 | FLAKY         | 3/6      |
</details>
<!--/body-->

rockwotj commented on src/v/iceberg/datatypes.h:
```diff
@@ -123,10 +123,30 @@ struct nested_field {
 
-    static nested_field_ptr
-    create(int32_t id, ss::sstring name, field_required req, field_type t) {
+    using is_new = ss::bool_class<struct new_field_tag>;
+    using removed = ss::bool_class<struct removed_field_tag>;
```
Are these always `true`? In this case should they just be empty structs to signify that? `struct is_new` and `struct removed`? If not, I don't really understand what it means for one of these to be `false`, because you can have is_new::no and removed::no and they both seem to represent the same state?

oleiman replied:
Ah, good point. `is_new` is always true, should be an empty struct. Both values of `removed` are used to annotate the source schema. Maybe logically some states could collapse, but I prefer the semantics of "old schema fields are removed or not removed" vs "new schema fields are new or have `src_info`"

rockwotj replied:
Can we capture some of this as comments here then? As is it's quite hard to derive the intended usage

oleiman replied:
yeah there should be inline comments here. accidental omission when I was doc'ing everything

oleiman replied:
done

rockwotj commented on src/v/iceberg/compatibility_types.h:
```diff
@@ -23,4 +26,41 @@ enum class compat_errc {
+    null_nested_field,
+    invalid_state,
+};
+std::ostream& operator<<(std::ostream&, const schema_evolution_errc&);
```
Where possible, we want to try to bias towards `fmt::formatter` instead of `operator<<`

oleiman replied:
done

rockwotj commented on src/v/iceberg/compatibility_types.h:
```diff
@@ -23,4 +26,41 @@ enum class compat_errc {
+    size_t n_removed{0};
+    size_t n_added{0};
+    size_t n_promoted{0};
+    friend schema_transform_state&
```
does this need to be a `friend`?

oleiman replied:
nope

oleiman replied:
done

rockwotj commented on src/v/iceberg/datatypes.h:
```diff
@@ -123,10 +123,30 @@ struct nested_field {
+        std::optional<primitive_type> type{};
+    };
+    using evolution_metadata = std::variant<src_info, is_new, removed>;
+    mutable std::optional<evolution_metadata> meta;
```
Wrapping a `std::variant` in a `std::optional` is a little awkward. Should we fold the `nullopt` case into the `variant directly`?

oleiman replied:
yeah good call

oleiman replied:
done

rockwotj commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -330,3 +330,609 @@ struct_type nested_test_struct() {
+            ids.get_one(),
+            "qux",
+            field_required::yes,
+            list_type::create(2, field_required::yes, struct_type{})));
```
why `2`? Should this be using the `ids` generator?

oleiman replied:
ya, typo

oleiman replied:
done

rockwotj commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -330,3 +330,609 @@ struct_type nested_test_struct() {
+      },
+  },
+  struct_evolution_test_case{
+    .description = "removing a required field works",
```
I thought we just make these `null` - this isn't backwards compatible... I guess the Iceberg spec allows this so it's fine?

oleiman replied:
> I thought we just make these `null`

Ah, that's the other thing we talked about last week. There's a product preference for "clean" removal of columns. I don't feel strongly about it either way, considering 👇 

> this isn't backwards compatible

I think it is. As a parquet reader, I assume a column whose ID doesn't appear in the schema is just skipped?

> Iceberg spec allows this

☝️

rockwotj replied:
> As a parquet reader, I assume a column whose ID doesn't appear in the schema is just skipped?

I'm talking if you have a query that is `SELECT foo, bar FROM table` and you delete column `bar` then the query fails.

rockwotj replied:
Some automation will list all of the column names in a SELECT statement, but maybe this isn't a big deal because someone will have to explicitly remove said column from source schema and this is a case of we recommend you turn on iceberg compat checks in SR first

oleiman replied:
> I'm talking if you have a query that is `SELECT foo, bar FROM table`

Interesting. I think we'll need to revisit this with @mattschumpert , as the whole motivation for dropping the column "for real" was something like "you dropped a field from your schema, it's confusing if it still appears there, even if always null".

I don't have a good sense of which invariant is more important from a product perspective. Doesn't seem like a blocker on this PR though.

rockwotj replied:
I think it might be OK if we drop something for real to mirror that. Connector based approaches don't do that AFAIK, but that might be something that can be done and we just recommend enabling the SR compat checks so you don't break queries

andrwng replied:
>Iceberg spec allows this

For my own understanding, can you elaborate on how this behaves with various query engines? Does this make the entire table unusable for certain query engines? Or does this only affect as-of queries? Or is it something that's inconsistent across engines at the moment?

Presumably if the current schema has a required field removed, subsequent queries should correctly not be able to use that field, though I guess that's only unsurprising if we think of "required" as "non-nullable"

rockwotj replied:
From a SQL statement/validation side of things removing columns is only an issue if you explicitly use that column. If you're just doing SELECT * or never referencing it explicitly then it's fine to remove.

> Presumably if the current schema has a required field removed, subsequent queries should correctly not be able to use that field, though I guess that's only unsurprising if we think of "required" as "non-nullable"

If you're saying the case of making a column nullable from NOT NULL, then yes nothing "breaks" in terms of the query runs, but predicates/transforms on that column may have unintended side effects, but they should not "break" from my understanding of ANSI SQL. If a query engine has different semantics we can't really do anything about it.

rockwotj replied:
Anyways, I think it's fine to remove columns that are removed from the source schema because we can add compat checks later to SR that would catch these (and likely this would be caught in the existing SR validation).

rockwotj commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -330,3 +330,609 @@ struct_type nested_test_struct() {
+
+          return all_updated && structs_equivalent(src, dst);
+      },
+    // TODO(oren): do we need to detect field reordering? or just support it?
+    // should reordered fields in a schema correspond to some parquet layout
+    // change on disk?
+    .any_change = schema_changed::no,
```
I think schema registry canonicalization should save us from some of this (we should probably check that we're traversing the protobuf schema correctly), but I don't think we should care about ordering of fields, since parquet is columar it should not matter the order of columns (where in row oriented formats you might be able to stop parsing early if you don't need all the fields?)

oleiman replied:
>schema registry canonicalization should save us

great point

> parquet is columnar it should not matter the order of columns

That's my read of the correctness rule - i.e. "it doesn't matter what order fields appear in the schema" vs "you can change the column order by schema evolution".

rockwotj reviewed (commented):
Can you lay out the next steps from this piece? Here we have the ability to validate and have all the delta information between the two schemas.

I'm assuming we need to:
* apply transformations in the data path from old data to new schemas based on the metadata here.
* alter tables as needed in iceberg (which there is already some code for right?)

rockwotj commented:
> Can you lay out the next steps from this piece? Here we have the ability to validate and have all the delta information between the two schemas.
> 
> I'm assuming we need to:
> 
>     * apply transformations in the data path from old data to new schemas based on the metadata here.
> 
>     * alter tables as needed in iceberg (which there is already some code for right?)

I think the first one is covered by `CORE-8530` right?

oleiman commented:
In my mind, both of these are covered by 8530.

>     * apply transformations in the data path from old data to new schemas based on the metadata here.

Or something. @andrwng had suggested another approach that I understood as (paraphrasing) "keep writing 'old data' to the original Parquet file with the understanding that it can be read by the new schema". So we need to agree on implementation detail, but generally keeping forward compatible records off the DLQ is the goal.

>     * alter tables as needed in iceberg (which there is already some code for right?)

Yup, though I think most of the existing code is fit for purpose. There's a bunch of stuff in place already for updating schemas, so we're mostly adding support for "interesting" updates.

rockwotj commented:
> but generally keeping forward compatible records off the DLQ is the goal.

Agreed on the end goal - I'm not going to fight too hard for one or the other, but note that you can have producers writing both, so you might get small files during a translation interval. It does seem simpler tho so I'm happy to go that route if that's true.

oleiman commented:
> > but generally keeping forward compatible records off the DLQ is the goal.
> 
> Agreed on the end goal - I'm not going to fight too hard for one or the other, but note that you can have producers writing both, so you might get small files during a translation interval. It does seem simpler tho so I'm happy to go that route if that's true.

Yeah no strong opinion on my end either, though my initial instinct was to upcast.

rockwotj commented:
In general this all LGTM outside the nits and small cleanups around the variant, I just want to make sure I have a good understanding of the current standing of things and the roadmap here. Thanks for this!

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/02b5d2c1c8652e12ca958fd9374df97505525bbd..3094cfeface5f9517d35588fc872ffa1fbca4cbe) CR comments (comments & adjusting the shape of `evolution_metadata` variant)

oleiman commented:
CI Failure:
- [build-release](https://buildkite.com/redpanda/redpanda/builds/60362#01944229-fe44-426b-9312-e120ca4c160d) I think was infra?

rockwotj commented:
Yeah I think a bunch of agents got nuked or something 🤷

oleiman commented:
/ci-repeat 1

andrwng commented on src/v/iceberg/datatypes.h:
```diff
@@ -123,10 +123,54 @@ struct nested_field {
-    static nested_field_ptr
-    create(int32_t id, ss::sstring name, field_required req, field_type t) {
+    /**
+     * evolution_metadata - Variant holding annotation related to schema
+     * evolution (i.e. transforming some 'source' schema into some 'destination'
+     * schema)
```
nit: it's a little unclear on a first pass whether in this description, the `this` `nested_field` is the expected to be source or the destination. Can you clarify that?

andrwng replied:
Made it through the PR, i guess it's either -- still I think it's worth calling out explicitly since it's not quite clear unless you start reading this code from the compatibility modules

oleiman replied:
no problem

oleiman replied:
done. lmk if it's unclear

andrwng commented on src/v/iceberg/compatibility_types.h:
```diff
@@ -23,4 +25,49 @@ enum class compat_errc {
 
+// TODO(oren): possibly a whole error_code?
+enum class schema_evolution_errc {
+    success,
```
nit: i know this is common practice for `error_code`, but if this isn't an error code and is meant to be used with a result class, it's worth considering not having a `success` case, if anything to encourage fully covered switches (though maybe this enum is large enough that it's unwieldy to do so)

oleiman replied:
actually this is an artifact from a previous iteration - only used in tests at this point as a placeholder and has no real application level semantics. will remove.

oleiman replied:
done

andrwng commented on src/v/iceberg/compatibility_utils.h:
```diff
@@ -0,0 +1,170 @@
+
+/**
+ * for_each_field - Apply some function to a field and all fields
+ * nested under it.
```
nit: can you add whether the function is run on fields breadth first or depth first?

oleiman replied:
done

andrwng commented on src/v/iceberg/datatypes.h:
```diff
@@ -123,10 +123,54 @@ struct nested_field {
+    };
+    struct is_new {};
+    using removed = ss::bool_class<struct removed_field_tag>;
+    using evolution_metadata
+      = std::variant<std::nullopt_t, src_info, is_new, removed>;
+    mutable evolution_metadata meta{std::nullopt};
```
I think I roughly understand why it is the case, but it's interesting that this field isn't used in the comparison operator of `nested_field`. Just want to call that out to make sure that's intentional.

It also makes me wonder whether we should do schema evolution on some  `struct field_and_evo_meta { field, evolution_metadata }` rather than on the fields themselves, though I guess that'd also require use to duplicate a ton of code for "evolvable complex types"..

oleiman replied:
Yeah, it's intentional. The general idea is to distinguish this little annotation field from those which carry Iceberg semantics.

As far as wrapping in an aggregate, I thought about that. I also have a previous iteration in the can that does the traversal and builds up a flat vector of "actions" (w/ field references inside) that can be evaluated _post facto_.

I think these alternatives mostly work fine (if they require more code), but it's a bit cumbersome to reason about pushing that information down towards table metadata w/o more API shenanigans.

I agree it looks sort of wacky, but my intuition is that sticking this extra couple bytes on each field is the simplest thing.

andrwng commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -325,3 +325,613 @@ struct_type nested_test_struct() {
+                 && updated(*src_map.value_field, *dst_map.value_field);
+      },
+  },
+  struct_evolution_test_case{
+    .description = "we can 'add' nested fields",
+    .generator = [](unique_id_generator&) { return struct_type{}; },
```
Is the case of adding fields in the middle of the schema allowed yet?

oleiman replied:
yes. if there's no test case for it I'll add one (or make this one more general).

oleiman replied:
done

andrwng commented on src/v/iceberg/compatibility.h:
```diff
@@ -40,4 +40,33 @@ namespace iceberg {
+ *
+ * @return std::nullopt (indicating success), or an error code
+ */
+schema_transform_result
+annotate_schema_transform(const struct_type& source, const struct_type& dest);
```
Just a thought, this seems like a thing we could fuzz test

oleiman replied:
yeah i agree. probably should. noted for a follow up 🙂

andrwng commented on src/v/iceberg/tests/compatibility_test.cc:
```diff
@@ -325,3 +325,613 @@ struct_type nested_test_struct() {
+    return os << tc.description;
+}
+
+static const std::vector<struct_evolution_test_case> valid_cases{
```
Wondering whether the current code supports multiple updates in the same evolution?

oleiman replied:
you mean prior to this diff? I think so

andrwng replied:
I meant with the new compatibility checks -- for instance, I was wondering if it made sense to test an add and remove in the same evolution (unless i missed cases where multiple operations are happening)

oleiman replied:
oh, yeah definitely. i don't think there's a specific test geared towards that, but the one whose description begins "grouping multiple fields into a struct..." has that effect

andrwng commented on src/v/iceberg/tests/BUILD:
```diff
@@ -649,3 +649,19 @@ redpanda_cc_bench(
         "@seastar//:benchmark",
     ],
 )
+
```
>iceberg/compat: Microbenchmark for evolve_schema & friends

nit: got any numbers? Might be worth posting some in the commit message. Or at least any conclusions that you've drawn

oleiman replied:
> posting some in the commit message

ah yeah, slipped my mind. thx

oleiman replied:
done

andrwng reviewed (commented):
Nice! Overall this looks pretty solid.

>"keep writing 'old data' to the original Parquet file with the understanding that it can be read by the new schema". So we need to agree on implementation detail

Yea the thought here was that it's hopefully uncommon for many schemas to be in use at once, and that with out current data primitives it's probably a good amount of non-trivial code to cast all the different kinds of schema changes to the actual data (especially thinking about field removal, nullability, and such).

My hunch is that the onus would generally be on query engines to reconcile the data schema with the current Iceberg schema. If that is the case, the "write old data" approach seems pretty reasonable. Wondering if either of you have had experience with that yet?

rockwotj commented:
> Nice! Overall this looks pretty solid.
> 
> > "keep writing 'old data' to the original Parquet file with the understanding that it can be read by the new schema". So we need to agree on implementation detail
> 
> Yea the thought here was that it's hopefully uncommon for many schemas to be in use at once, and that with out current data primitives it's probably a good amount of non-trivial code to cast all the different kinds of schema changes to the actual data (especially thinking about field removal, nullability, and such).
> 
> My hunch is that the onus would generally be on query engines to reconcile the data schema with the current Iceberg schema. If that is the case, the "write old data" approach seems pretty reasonable. Wondering if either of you have had experience with that yet?

No I actually agree here that this is the simpler path forward and we should start with this. With compaction and the assumption that schema evolution is not a frequent operation, I think we may not need the more complicated thing.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60369-->
<!--3c444ac7-858e-4801-b74c-eab0fea8e976-->
<!--title-->
## Retry command for [Build#60369](https://buildkite.com/redpanda/redpanda/builds/60369)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/consumer_offsets_consistency_test.py::ConsumerOffsetsConsistencyTest.test_flipping_leadership
```
<!--/body-->

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/3094cfeface5f9517d35588fc872ffa1fbca4cbe..1bf54b78a4e19b816c2d69fca2e878aa96246b8b) contents (CR):
- removed an unnecessary errc case
- fixed up some doc comments
- added a test case

andrwng commented on src/v/iceberg/compatibility_utils.h:
```diff
@@ -0,0 +1,170 @@
+  S& s, detail::FieldOp<T*> auto&& fn, detail::FieldPredicate auto&& filter) {
+    static_assert(std::is_const_v<S> == std::is_const_v<T>);
+    chunked_vector<T*> stk;
+    for (const auto& f : std::ranges::reverse_view(s.fields)) {
```
Interesting that this was reversed between revisions without too much recourse. Curious what the implication is/was?

oleiman replied:
Nothing functional. Just an accidental omission - I intended to duplicate the control flow of the field collector visitor, though it shouldn't have any effect here.
